x-pack/plugin/esql/src/main/java/org/elasticsearch/xpack/esql/expression/function/scalar/hash/Hash.java

```diff
@@ -0,0 +1,115 @@
+import org.elasticsearch.compute.ann.Evaluator;
+import org.elasticsearch.compute.ann.Fixed;
+import org.elasticsearch.compute.operator.EvalOperator;
+import org.elasticsearch.xpack.esql.core.expression.Expression;
+import org.elasticsearch.xpack.esql.core.tree.NodeInfo;
+import org.elasticsearch.xpack.esql.core.tree.Source;
+import org.elasticsearch.xpack.esql.core.type.DataType;
+import org.elasticsearch.xpack.esql.expression.function.FunctionInfo;
+import org.elasticsearch.xpack.esql.expression.function.Param;
+import org.elasticsearch.xpack.esql.expression.function.scalar.EsqlScalarFunction;
+import org.elasticsearch.xpack.esql.io.stream.PlanStreamInput;
+
+import java.io.IOException;
+import java.security.MessageDigest;
+import java.security.NoSuchAlgorithmException;
+import java.util.HexFormat;
+import java.util.List;
+
+public class Hash extends EsqlScalarFunction {
+
+    public static final NamedWriteableRegistry.Entry ENTRY = new NamedWriteableRegistry.Entry(Expression.class, "Hash", Hash::new);
+
+    private final Expression alg;
+    private final Expression input;
+
+    @FunctionInfo(returnType = "keyword", description = "Computes the hash of the input using the supplied algorithm.")
+    public Hash(
+        Source source,
+        @Param(name = "alg", type = { "keyword", "text" }, description = "Hash algorithm to use.") Expression alg,
+        @Param(name = "input", type = { "keyword", "text" }, description = "Input to hash.") Expression input
+    ) {
+        super(source, List.of(alg, input));
+        this.alg = alg;
+        this.input = input;
+    }
+
+    private Hash(StreamInput in) throws IOException {
+        this(Source.readFrom((PlanStreamInput) in), in.readNamedWriteable(Expression.class), in.readNamedWriteable(Expression.class));
+    }
+
+    @Override
+    public void writeTo(StreamOutput out) throws IOException {
+        source().writeTo(out);
+        out.writeNamedWriteable(alg);
+        out.writeNamedWriteable(input);
+    }
+
+    @Override
+    public String getWriteableName() {
+        return ENTRY.name;
+    }
+
+    @Override
+    public DataType dataType() {
+        return DataType.KEYWORD;
+    }
+
+    @Override
+    public boolean foldable() {
+        return alg.foldable() && input.foldable();
+    }
+
+    @Evaluator(warnExceptions = NoSuchAlgorithmException.class)
+    static BytesRef process(BytesRef alg, BytesRef input) throws NoSuchAlgorithmException {
+        return hash(MessageDigest.getInstance(alg.utf8ToString()), input);
+    }
+
+    @Evaluator(extraName = "Constant")
+    static BytesRef processConstant(@Fixed(build = true) MessageDigest alg, BytesRef input) {
+        return hash(alg, input);
+    }
+
+    private static BytesRef hash(MessageDigest alg, BytesRef input) {
+        alg.update(input.bytes, input.offset, input.length);
+        var result = alg.digest();
+        return new BytesRef(HexFormat.of().formatHex(result));
+    }
+
```

Comment:
I think this wants a `resolveType` method that forces the parameters to be strings. Without that folks can do something like
```
ROW a = HASH("md5", 12)
```

and we'll end up with runtime exceptions casting stuff to `BytesRefBlock`. The compute engine wants an external system - like the language itself - to make sure to only build trees of operators that make valid types. And `resolveTypes` is the trick that does it.

Reply:
Also things like:
```
ROW a = HASH("md5", "192.168.0.1"::IP)
```

will work. But it'll hash the 128 bit representation we use for IPs. That might even be what people want, but it's maybe unexpected.

Reply:
To clarify the comment above : "192.168.0.1" as a string gets converted into an IP. Which then gets automatically converted into a byteref; please check whether it's the original "192.168.0.1" or something else.